src/util/expr_cast.h

```diff
@@ -0,0 +1,157 @@
+#ifndef CPROVER_UTIL_EXPR_CAST_H
+#define CPROVER_UTIL_EXPR_CAST_H
+
+/*!
+ * \file util/expr_cast.h
+ * \brief Templated functions to cast to specific exprt-derived classes
+ */
+
+/*
+  Author: Nathan Phillips <Nathan.Phillips@diffblue.com>
+*/
+
+#include <typeinfo>
+#include "invariant.h"
+#include "expr.h"
+
+
+/*!
+ * \brief Check whether a reference to a generic \ref exprt is of a specific derived class
+ *   Implement template specializations of this function to enable casting
+ * \tparam T The exprt-derived class to check for
+ * \param base Reference to a generic \ref exprt
+ * \return true if \a base is of type \a T
+ */
+template<typename T> bool check_expr_type(const exprt &base);
+
+/*!
+ * \brief Check whether a reference to a generic \ref exprt is of a specific derived class
+ *   Implement template specializations of this function to enable casting
+ * \tparam T The exprt-derived class to check for
+ * \param base Reference to a generic \ref exprt
+ * \return true if \a base is of type \a T
+ */
+template<typename T> void validate_expr(const T &value) { }
```

Comment:
I think plain function overloading would be better than specialization. Also, I'd consider combining `check_expr_type` and `validate_expr` seeing as their roles are so similar. If we *just* used `validate_expr`, we could ignore template specialization completely, and simply provide overloads for all the different types we want to support.